telink: b92: fix partial flash protect .

haiwentelink opened the pull request:
- change the flash protect to adpot partial protect.

andriy-bilynskyy commented:
The following west manifest projects have been modified in this Pull Request:

| Name | Old Revision | New Revision | Diff |
| ---- | ------------ | ------------ |------|
| hal_telink | https://github.com/telink-semi/hal_telink/commit/8e22fad3c938f4ea2266293897d8b8ea47e566a0 | https://github.com/telink-semi/hal_telink/commit/34700f71d3c69326595484c631853ca61fabf0af (develop) | [telink-semi/hal_telink@8e22fad3..34700f71](https://github.com/telink-semi/hal_telink/compare/8e22fad3c938f4ea2266293897d8b8ea47e566a0..34700f71d3c69326595484c631853ca61fabf0af) |

*Note: This message is automatically posted and updated by the Manifest GitHub Action.*

andriy-bilynskyy commented on drivers/flash/soc_flash_b9x.c:
```diff
@@ -105,31 +117,39 @@ static void flash_b9x_modify(struct flash_b9x_data *dev_data, uintptr_t offset,
+			uint8_t *p_sector = malloc(SECTOR_SIZE);
+
+			if (p_sector == NULL) {
+				return;
```
Please add LOG_ERR here

haiwentelink replied:
will add in this func .

andriy-bilynskyy commented on drivers/flash/soc_flash_b9x.c:
```diff
@@ -105,31 +117,39 @@ static void flash_b9x_modify(struct flash_b9x_data *dev_data, uintptr_t offset,
 	while (len) {
 		size_t len_page_end =  SECTOR_SIZE - off_sector;
 		size_t len_current = len_page_end;
+		bool clean_flag = true;
```
No needed to assign this value.

haiwentelink replied:
will change

andriy-bilynskyy commented on drivers/flash/soc_flash_b9x.c:
```diff
@@ -37,22 +37,24 @@ static const struct flash_parameters flash_b9x_parameters = {
 }
 
-static void flash_b9x_lock_start(void)
+static unsigned int lock_offset;
```
Static variableas are not allowed in drivers. Each device has own data and configuration.

haiwentelink replied:
can we assign the adr into the struct of the flash_b9x_parameters

andriy-bilynskyy commented on drivers/flash/soc_flash_b9x.c:
```diff
@@ -37,22 +37,24 @@ static const struct flash_parameters flash_b9x_parameters = {
 }
 
-static void flash_b9x_lock_start(void)
+static unsigned int lock_offset;
```
What this offset is ding?

haiwentelink replied:
we want to transfer the offset address to lock func , but the lock fun will not run immidately ,if not keep , the value will lost .

andriy-bilynskyy commented on drivers/flash/soc_flash_b9x.c:
```diff
@@ -346,17 +366,9 @@ static const struct flash_driver_api flash_b9x_api = {
-										    \
 	DEVICE_DT_INST_DEFINE(n, flash_b9x_init,			\
 			      NULL,				    \
-			      &flash_data_##n,				    \
```
That's Zephyr drivers convention to use device data not static variables.
It solves the problem with multiply instances.

haiwentelink replied:
if don't use the flash_data, I think this value can set to be null .

andriy-bilynskyy commented on drivers/flash/soc_flash_b9x.c:
```diff
@@ -15,6 +15,7 @@
 #include <zephyr/device.h>
 #include <zephyr/drivers/flash.h>
 #include <zephyr/kernel.h>
+#include <stdlib.h>
```
The driver is already using k_malloc which allocates data in system heap not user.
For now it's using both heaps.
Clear way is to use only system heap for drivers. But if we want to have memory pofit lets use user in all driver code, anyway we have no MPU existed/setuped.

haiwentelink replied:
the init config heap is configed to 512, but the heap of the lib is very large(12k) , so I think there is no need to enlarge the system heap , multi use of the heap .

haiwentelink commented:
@andriy-bilynskyy  can start review again , delete the timer for lock and unlock , can call directly , for it will not excute frequently . and fix the bugs of the pwm driver for the interface changed .

andriy-bilynskyy commented:
Plese update HAL Telink 34700f71d3c69326595484c631853ca61fabf0af

haiwentelink committed:
telink: flash: fix flash and pwm driver.

- use malloc to instead of the static buffer .
- change the flash protect to adpot partial protect.
- fix pwm driver for the interface changes.

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>
